typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,996 @@
+
+Behold, the lambda cube:
+
+<center>
+
+![An image of the lambda cube](./lambda-cube-plain.svg)
+
+</center>
+
+Going directly up in the picture introduces dependence of types upon types, i.e. user-defined type constructors.
+Going left introduces dependence of terms upon types, i.e. System F style polymorphism.
+Going right introduces dependence of types upon terms, i.e. Martin-Löf style dependent types.
+
+Thus, we have the following variations:
+
+| | term ← type | type ← type | type ← term |
+|-------------------------------|:-:|:-:|:-:|
+| $\lambda_\to$                 |   |   |   |
+| $\lambda_2$                   | × |   |   |
+| $\lambda_{\underline\omega}$  |   | × |   |
+| $\lambda_P$                   |   |   | × |
+| $\lambda_\omega$              | × | × |   |
+| $\lambda_{P2}$                | × |   | × |
+| $\lambda_{P\underline\omega}$ |   | × | × |
+| $\lambda_C$                   | × | × | × |
+
+The left arrow ← denotes "depends on".
+
+The lambda cube was introduced by Henk Barendregt in (Barendregt, 1991) to investigate different generalizations of simply-typed lambda calculus (here denoted as $\lambda_\to$).
+
+Let us briefly discuss some of the options.
+
+$\lambda_2$ is System F, discussed above.
+
+$\lambda_\omega$ is System Fω.
+It extends System F with type operators, and is a different, much more complex system.
+In short, System Fω allows defining type constructors inside the calculus itself instead of considering those as primitives, i.e. allows for user-defined types.
+
+$\lambda_{\underline\omega}$, a.k.a. System F<u>ω</u> is System Fω without type abstraction/application, or simply typed lambda calculus with user-defined types.
+
+$\lambda_P$ is simply typed lambda calculus with dependent types.
+
+$\lambda_C$, a.k.a. $\lambda_{P\omega},$ is the aforementioned calculus of constructions (Coquand, 1986), famously used as the basis for the Coq proof assistant, but also with other dependently-typed languages.
+Here, the border between terms and types is virtually gone.
+
+We will not discuss these advanced type systems in any more detail here, but I thought I should at least mention them.
+
+## Wadler-Blott type classes
+
+What about ad-hoc polymorphism then?
+Can it be introduced into typed lambda calculi in a more or less principled fashion?
+The answer, as it turns out, is yes.
+This is shown by Philip Wadler and Stephen Blott in (Wadler, 1989a).
+
+The idea here is to make polymorphism _constrained_.
+That is, instead of simply a $\forall \alpha$ quantifier, we may have a $\forall \alpha. P\;\alpha,$ where $P$ is some predicate.
+
+With type classes specifically, predicate $P$ can essentially only be "does this type belong to an open class of types".
+
+A type class defines an "interface", a set of functions that can be used with any type in the class, but the class itself usually provides no implementation.
```

Comment:
Is it possible to put a link to our Haskell typeclass article somewhere here? (https://serokell.io/blog/haskell-typeclasses) Or would it not be good?